teps/0074-deprecate-pipelineresources.md

```diff
@@ -0,0 +1,366 @@
+  * User-definable Steps? This is what Tasks provide.
+  * User-definable params? Tasks already have these.
+  * User-definable "resource results"? Tasks have Task Results.
+  * Sharing data between Tasks using PVCs? workspaces provide this for Tasks.
+
+So the choice is: do we fix PipelineResources such that the above issues are mitigated, or do we remove them completely?
+This proposal suggests that we remove them completely. A similar concept could still be added in the future, but it
+could be designed from the start to avoid the problems we've run into, and meanwhile we can go ahead with our v1 plans
+without having to solve this abstraction first (also the fact that this hasn't been fixed yet is a possible indication
+that maybe this concept, at least in its current form, isn't required).
+
+### Goals
+
+* Get ready for [the v1 Tekton Pipelines release](https://github.com/tektoncd/pipeline/issues/3548) by either deciding
+  to bring PipelineResources to beta and then v1, or removing them
+* Finally address PipelineResources which have been languishing at alpha since our
+  [Tekton Pipelines beta release in Mar 2020](https://github.com/tektoncd/pipeline/releases/tag/v0.11.0) but are
+  [still confusingly part of these beta APIs](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#specifying-resources)
+  (that is, beta Tasks and Pipelines are still coupled to these alpha types) in their APIs.
+
+### Non-Goals
+
+* State absolutely that a similar concept is not useful and should not be explored in the future: similar ideas have
+  been popping up in different places, for example:
+  * The pipelines-as-code [Repository CRD](https://github.com/openshift-pipelines/pipelines-as-code#features)
+  * Not explicitly mentioned in the proposal but the concept has been discussed in the context of the
+    [Tekton workflows feature](https://github.com/tektoncd/community/issues/464)
+  * [On twitter](https://twitter.com/bobcatwilson/status/1408064403585441804)
+
+### Use Cases
+
+* As a user of Tekton Pipelines v1, I want to be able to:
+  * Understand clearly the stability level of all components I'm using
+  * Know which components I should depend on and are likely to continue to be supported in the future
+    * I want this to be clear in the documentation, examplesa, and dogfooding done on the project itself
+* As a maintainer of Tekton Pipelines, I want to know if I should continue to fix bugs with and add features to
+  PipelineResources
+
+## Requirements
+
+* When we create v1 versions of Tasks, Pipelines, TaskRuns and PipelineRuns, all alpha features should be behind 
+  [the alpha api flag](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#customizing-the-pipelines-controller-behavior)
+* Our documentation and examples should clearly recommend best practices for how to use Tekton Pipelines components
+* Our dogfooding should provide guidelines for how we recommend using Tekton
```

Comment:
Do we want to have as requirements that any TEP that is listed above (and implementing parts of what PipelineResource was providing) should be implemented and out of alpha ? (aka enabled by default)

Reply:
That's a good question - would you want that to be a requirement before we announce our intention to deprecate PipelineResources, or would that be a requirement for their final removal?

1. One path could be that we approve this TEP as "implementable", announce the deprecation, add it to our v1 plans, but we make their actual removal (and therefore the v1 release!) contingent on these other TEPs making it to at least alpha. (Perhaps we include that the TEPs should at least be accepted as proposed? which TEP-0044 (decoupling scheduling) is, but the other two (dict + array support) are not)
2. The other path would be to maintain our current state (PipelineResources not recommended but with their future somewhat ambiguous) until those other TEPs make it to at least Alpha

I feel like (1) has the disadvantage of giving our users somewhat conflicting info: this thing is deprecated, the replacement(s) are a WIP, but (2) has the disadvantage of continuing this limbo state where we're not giving users a clear idea of our plans. So my preference (and there are definitely some selfish reasons here but I'll try to be impartial XD) would be to go with (1)

Reply:
+1 to option (1) - if we know this is the direction we want to go it seems reasonable to warn users ASAP so they don't adopt something we plan to remove in the near future (even if the alternative isn't as available / as friendly yet).

Reply:
option (1) seems more appropriate indeed, and announcing the `PipelineResources` are deprecated is more or less what we unofficially say (at least I do 😅 ) so this would make it official. *But* I think we should aim for condition the "removing them" (CRD not in new release / fresh installs) to "having any feature required to support the use case PipelineResource were out of alpha"  (aka available on a default install) — *what weird sentence lol*. If we don't have that, indeed the message is going to be weird : "you could do that before, now you have to enable alpha api to do the same".

So I vote for marking this implementable, announcing the deprecation, but adding a condition to the actual removal, which would be "the *required* set of feature (list of TEP) is implemented, and out-of beta for X release".